docs/50_writing_tests.md

```diff
@@ -0,0 +1,463 @@
+            bluetoothDeviceInfos.push_back(hasInvalid);
+            bluetoothDeviceInfos.push_back(hasBoth);
+        }
+    }
+
+public:
+    StagesBike3TestData() : StagesBikeTestData("Stages Bike (KICKR CORE)") {
+
+        this->addDeviceName("KICKR CORE", comparison::StartsWithIgnoreCase);
+    }
+};
+```
+
+In this case, it populates the vector with the single enabling configuration if that's what's been requested, otherwise 3 disabling ones.
+
+### Exclusions
+
+Sometimes there might be ambiguity when multiple devices are available, and the detection code may specify that if the other conditions match, but certain specific kinds of devices (the exclusion devices) have already been detected, the newly matched device should be ignored.
+
+The TestData class can be made to cover this by overriding the configureExclusions() method to add instances of the TestData classes for the exclusion devices to the object's internal list of exclusions.
+
+Detection code:
+
+```
+} else if (b.name().startsWith(QStringLiteral("ECH")) && !echelonRower && !echelonStride &&
+                       !echelonConnectSport && filter) {
+```
+The configureExclusions code is overridden to specify the exclusion test data objects. Note that the test for a previously detected device of the same type is not included.
+
+```
+#pragma once
+
+#include "Devices/Bike/biketestdata.h"
+#include "Devices/EchelonRower/echelonrowertestdata.h"
+#include "Devices/EchelonStrideTreadmill/echelonstridetreadmilltestdata.h"
+#include "devices/echelonconnectsport/echelonconnectsport.h"
+
+class EchelonConnectSportBikeTestData : public BikeTestData {
+
+public:
+    EchelonConnectSportBikeTestData() : BikeTestData("Echelon Connect Sport Bike") {
+        this->addDeviceName("ECH", comparison::StartsWith);
+    }
+
+    void configureExclusions() override {
+        this->exclude(new EchelonRowerTestData());
+        this->exclude(new EchelonStrideTreadmillTestData());
+    }
+
+    deviceType get_expectedDeviceType() const override { return deviceType::EchelonConnectSport; }
+
+    bool get_isExpectedDevice(bluetoothdevice * detectedDevice) const override {
+        return dynamic_cast<echelonconnectsport*>(detectedDevice)!=nullptr;
+    }
+};
+
+
+```
+
+### When a single TestData Class Can't Cover all the Conditions
+
+Detection code:
+
+```
+QString powerSensorName =
+        settings.value(QZSettings::power_sensor_name, QZSettings::default_power_sensor_name).toString();
+...
+} else if ((b.name().toUpper().startsWith(QStringLiteral("STAGES ")) ||
+                        (b.name().toUpper().startsWith("TACX SATORI")) ||
+                        ((b.name().toUpper().startsWith("KICKR CORE")) && !deviceHasService(b, QBluetoothUuid((quint16)0x1826)) && deviceHasService(b, QBluetoothUuid((quint16)0x1818))) ||
+                        (b.name().toUpper()==QStringLiteral("QD")) ||
+                        (b.name().toUpper().startsWith(QStringLiteral("ASSIOMA")) &&
```

Comment:
https://github.com/cagnulein/qdomyos-zwift/blob/2834f6c5d816a93ac591f736363ebb233c07916f/src/devices/bluetooth.cpp#L1452

@cagnulein I have changed the code in this example. In bluetooth.cs it's checking that it starts with QD and is 2 characters long. I wonder if that should just compare the upper cased strings with == or a length greater than 2 is expected.

Reply:
For clarity, I changed the code in the documentation example, not  bluetooth.cs.

Reply:
@drmason789 do you mean to scrictly checking the QD string? I have to blame that line to check how was the original device

Reply:
> For clarity, I changed the code in the documentation example, not bluetooth.cs.

yes  got it!

Reply:
> @drmason789 do you mean to scrictly checking the QD string? I have to blame that line to check how was the original device

I mean:

```
 (b.name().toUpper()==QStringLiteral("QD")) ||
```

Reply:
sorry it's not clear to me @drmason789 
the code on master is

`(b.name().toUpper().startsWith(QStringLiteral("QD")) && b.name().length() == 2) ||`

what's wrong with this?

Reply:
It's not wrong, but it's doing a starts with check and a length check. Because the length of the text it is supposed to start with is equal to the length the entire string is supposed to be, it can only ever accept "QD" so why not just use the equality operator instead?